### PR TITLE
feat: Add support for native v0 custom elements to behave like v1.

### PR DESCRIPTION
**Pull request by treshugart**
BREAKING CHANGE: v0 and v1 both allow whitespace of all kinds, so this has been relaxed. Implements
#108.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -137,26 +139,7 @@ function getSlotNameFromNode(node) {
 }
 
 function slotNodeIntoSlot(slot, node, insertBefore) {
-  // Don't slot nodes that have content but are only whitespace. This is an
```
Comment (treshugart) [inline src/index.js:140]: Unfortunately both v0 and v1 allow this, so I've relaxed this check to allow whitespace of all sorts. This just means that if we want fallback content to display that we need to ensure there's no unnecessary whitespace in our DOM.

File: src/index.js [hunk 2]
```diff
@@ -838,7 +821,11 @@ const members = {
   },
 };
 
-if (!('attachShadow' in document.createElement('div'))) {
+if (shadowDomV1) {
+  // then we should probably not be loading this
+} else if (shadowDomV0) {
+  v0();
+} else {
   const commProto = Comment.prototype;
```
Comment (treshugart) [inline src/index.js:829]: Eventually I'd like to do the same thing here that I did with `v0()` to simplify this file.

File: src/v0/index.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import 'custom-event-polyfill';
```
Comment (treshugart) [inline src/v0/index.js:1]: This is the meat of the code that enables native v0 to behave like v1.

File: test/unit/dom/childElementCount.js [hunk 4]
```diff
@@ -38,67 +36,41 @@ describe('dom: childElementCount', () => {
       it('should return correct number of child nodes', () => {
         numbers.forEach(num => {
           expect(elem.childElementCount).to.equal(num);
-          if (type !== 'host' && canPatchNativeAccessors) {
```
Comment (treshugart) [inline test/unit/dom/childElementCount.js:41]: Most of the changes to the tests involve removing code like this. While it was definitely thorough to test, every single one of these would break Chrome because we're depending on double-underscored members to be available. I don't think the cost-benefit ratio here is worth testing these as we should be getting expected values by testing the public API.

File: test/unit/dom/childNodes.js [hunk 5]
```diff
@@ -113,44 +85,26 @@ describe('dom: childNodes', () => {
...
 
         elem.appendChild(node3);
         expect(elem.childNodes[2]).to.equal(node3);
         expect(elem.innerHTML).to.equal('<node1></node1><node2></node2>text1');
-        if (canPatchNativeAccessors && type !== 'host') {
-          expect(elem.__childNodes[2]).to.equal(node3);
-        }
 
-        elem.insertBefore(node4);
+        elem.insertBefore(node4, null);
```
Comment (treshugart) [inline test/unit/dom/childNodes.js:97]: Native `insertBefore()` complains if `arguments.length !== 2`.

File: test/unit/dom/outerHTML.js [hunk 6]
```diff
@@ -1,6 +1,18 @@
-import canPatchNativeAccessors from '../../../src/util/can-patch-native-accessors';
+import { shadowDomV0 } from '../../../src/util/support';
 
 describe('dom: outerHTML', () => {
+  // This is so that it goes through any possible innerHTML massaging. For
+  // example, this is necessary when upgrading native v0 to behave like v1.
+  function html(str) {
```
Comment (treshugart) [inline test/unit/dom/outerHTML.js:6]: This exists so that comparison strings convert `<slot>` to `<content>`.

File: test/unit/shadow/polyfill.js [hunk 7]
```diff
@@ -27,20 +28,20 @@ describe('shadow/polyfill', () => {
     expect(host.shadowRoot).to.equal(null);
   });
 
-  it('polyfillShadowRootTagName: [default="_shadow_root_"]', () => {
+  shadowDomV0 || it('polyfillShadowRootTagName: [default="_shadow_root_"]', () => {
```
Comment (treshugart) [inline test/unit/shadow/polyfill.js]: These tests only apply to v1.

File: test/unit/slot/distribution.js [hunk 8]
```diff
@@ -56,23 +56,24 @@ describe('slot/distribution', () => {
...
-      host2.appendChild(slot1);
-      host2.attachShadow({ mode: 'open' });
-      host2.shadowRoot.appendChild(slot2);
+      // host1's slot1. Order of setup doesn't matter here.
 
       host1.appendChild(test);
       host1.attachShadow({ mode: 'open' });
       host1.shadowRoot.appendChild(host2);
 
+      host2.appendChild(slot1);
```
Comment (treshugart) [inline test/unit/slot/distribution.js:73]: I'm happy to reorder these, but when looking at it, reversing the setup here simplified the code while reading it.

File: test/unit/slot/fallback-content.js [hunk 9]
```diff
@@ -6,6 +7,11 @@ describe('fallback-content', () => {
   let slot;
   let fallback;
 
+  // We don't need to run these tests at all against v0.
```
Comment (treshugart) [inline test/unit/slot/fallback-content.js:10]: V0 fallback content behaves just like v1, so we don't have to do anything here.

File: test/unit/slot/polyfill.js [hunk 10]
```diff
@@ -95,8 +95,10 @@ describe('slot/polyfill', () => {
     host.appendChild(document.createTextNode(' '));
     host.appendChild(document.createTextNode('\n'));
     host.appendChild(document.createTextNode('testing'));
-    expect(slot2.assignedNodes().length).to.equal(2);
+    expect(slot2.assignedNodes().length).to.equal(4);
```
Comment (treshugart) [inline test/unit/slot/polyfill.js:98]: This is because we now allow all whitespace nodes.

File: src/util/support.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+const div = document.createElement('div');
+export const shadowDomV0 = !!div.createShadowRoot;
```
Comment (adevnadia) [inline src/util/support.js:2]: What about checking if this is a native function? It seems like everyone is messing with prototypes these days.

**Comment by adevnadia** [comment]
linting: ✖ 26 problems (18 errors, 8 warnings)

File: test/unit/dom/childElementCount.js (hunk 4 above)
Comment (adevnadia) [inline test/unit/dom/childElementCount.js:41]: Disagree about public API

**Comment by treshugart** [comment]
Must need to re-npm install to get updated linting stuff for my ide? Will try.

File: test/unit/dom/innerHTML.js [hunk 12]
```diff
@@ -33,10 +32,10 @@ describe('dom: innerHTML', () => {
           { value: 'text before <div>text inside</div> text after', nodes: 3, assigned: 3 },
           { value: 'text before <div>text inside <div>text deep inside</div> </div> text after', nodes: 3, assigned: 3 },
           { value: '<div><!-- comment inside --></div>', nodes: 1, assigned: 1 },
-          { value: '<!-- comment outside --> <div></div>', nodes: 3, assigned: 1 }, // empty space is a text node, do not slot empty space
-          { value: '<!-- comment outside --> <div><!-- comment inside --></div>', nodes: 3, assigned: 1 }, // empty space is a text node, do not slot empty space
-          { value: '<!-- comment before --> <div><!-- comment inside --></div> <!-- comment after -->', nodes: 5, assigned: 1 }, // empty space is a text node, do not slot empty space
-          { value: '<!-- comment before --> <div><!-- comment inside --><div><!-- comment deep inside --></div> </div><!-- comment after -->', nodes: 4, assigned: 1 }, // empty space is a text node, do not slot empty space
+          { value: '<!-- comment outside --> <div></div>', nodes: 3, assigned: 2 }, // empty space is a text node, do not slot empty space
```
Comment (adevnadia) [inline test/unit/dom/innerHTML.js:35]: Probably comments about 'do not slot empty space' not needed anymore?

File: src/util/support.js (hunk 11 above)
Comment (treshugart) [inline src/util/support.js:2]: That's shouldn't be necessary because if there's a v0 polyfill, then this should work. If not, that's a problem with their polyfill.

File: test/unit/dom/outerHTML.js [hunk 13]
```diff
@@ -37,118 +45,52 @@ describe('dom: outerHTML', () => {
...
-          if (canPatchNativeAccessors) {
-            expect(elem.__innerHTML).to.equal('<div4></div4>');
-          }
-        }
-
-        if (type === 'root') {
-          elem.innerHTML = '<div></div>';
-          // we can't change root with this, so nothing should happen
-          elem.outerHTML = '<p></p>';
-          expect(elem.outerHTML).to.equal('<_shadow_root_><div></div></_shadow_root_>');
```
Comment (adevnadia) [inline test/unit/dom/outerHTML.js:141]: Sure about deleting this test?

File: src/v0/index.js [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+
+  // Node
+  // ----
+
+  Object.defineProperty(Node.prototype, 'assignedSlot', {
+    get() {
+      const { parentNode } = this;
+      if (parentNode) {
+        const { shadowRoot } = parentNode;
```
Comment (adevnadia) [inline src/v0/index.js:100]: You assume that slot always has root as a parent? What about this situation: `<root><div><slot></slot></div></root>`?

**Comment by adevnadia** [comment]
Lots of new code, but no new tests at all?

File: test/unit/slot/fallback-content.js (hunk 9 above)
Comment (adevnadia) [inline test/unit/slot/fallback-content.js:10]: Shouldn't you prove it with tests?

**Comment by treshugart** [comment]
@nadiam84 I knew you'd say something ;)

We shouldn't need to test chromes native behaviour, only that it conforms to the v1 API, which we're already testing when it is run through our suite.

File: src/v0/index.js (hunk 14 above)
Comment (treshugart) [inline src/v0/index.js:100]: This would be a standard node. If it is slot assignable (light DOM) it's parent will be the host and should have a `shadowRoot` property. Shadow roots cannot be detached from a host and slots cannot work when outside of a shadow root. If the `shadowRoot` property doesn't exist, it means that it's "closed" and it should return null.

File: src/v0/index.js [hunk 15]
```diff
@@ -152,8 +152,8 @@ export default () => {
         const slots = {};
         const recordSlots = node => (slots[node.getAttribute && node.getAttribute('slot') || '__default'] = true);
 
-        addedNodes.forEach(recordSlots);
-        removedNodes.forEach(recordSlots);
+        [...addedNodes].forEach(recordSlots);
```
Comment (treshugart) [inline src/v0/index.js]: The only issue I see with this is that it can be expensive because this will do `Array.from()` in ES2015 natively and `Array.prototype.slice.call()` in ES5 (which is very expensive in performance critical code.

We might be able to get away with `Array.prototype.forEach.call(addedNodes)`. Failing that just putting an `if (addedNodesOrRemovedNOdes)` check around these would suffice.

Comment (treshugart) [inline src/v0/index.js]: That said, [there's still some issues](https://toddmotto.com/ditch-the-array-foreach-call-nodelist-hack/#problem-4-creates-a-needless-array) with that approach. I think the archaic `if` method is the best choice considering performance is probably the most important thing here.

File: src/index.js [hunk 16]
```diff
@@ -862,8 +849,14 @@ if (!('attachShadow' in document.createElement('div'))) {
       const nativeDescriptor = getPropertyDescriptor(elementProto, memberName);
       const nativeTextDescriptor = getPropertyDescriptor(textProto, memberName);
       const nativeCommDescriptor = getPropertyDescriptor(commProto, memberName);
-      const shouldOverrideInTextNode = (memberName in textNode && doNotOverridePropertiesInTextNodes.indexOf(memberName) === -1) || ~defineInTextNodes.indexOf(memberName);
-      const shouldOverrideInCommentNode = (memberName in commNode && doNotOverridePropertiesInCommNodes.indexOf(memberName) === -1) || ~defineInCommNodes.indexOf(memberName);
+      const shouldOverrideInTextNode = (
```
Comment (adevnadia) [inline src/index.js:852]: weird formating...

**Comment by adevnadia** [comment]
lgtm

File: src/index.js (hunk 16 above)
Comment (treshugart) [inline src/index.js:852]: agree but airbnb

File: test/unit/slot/fallback-content.js (hunk 9 above)
Comment (treshugart) [inline test/unit/slot/fallback-content.js:10]: Then we'd be testing chrome native. If there are tests missing for the polyfill, these should go into the slot polyfill tests and only be tested against the polyfill.

File: test/unit/dom/outerHTML.js (hunk 13 above)
Comment (treshugart) [inline test/unit/dom/outerHTML.js:141]: In its current state, yeah. It's far too confusing because we have 3 different environments:
1. Chrome native
2. WebKit
3. Standards compliant browsers that don't support v1 or v0

Comment (treshugart) [inline test/unit/dom/outerHTML.js:141]: To elaborate, the test description "should set a proper value" is kinda vague; I really don't know what it's testing here. We should isolate this down to a specific breakage in a given environment and test that only in that environment. Maybe we can pair on this at some point or raise an issue describing the exact problem so that we can implement it as a standalone test?

Comment (adevnadia) [inline test/unit/dom/outerHTML.js:141]: It just tests that 'set' for the outerHTML works as expected. It's a bit weird looking, I agree.

Comment (treshugart) [inline test/unit/dom/outerHTML.js:141]: I guess I'm just wondering if there's a specific issue we're checking for here. There must be, the test just doesn't say what it is.

File: src/index.js (hunk 16 above)
Comment (lukebatchelor) [inline src/index.js:852]: Is the `~defineInCommNodes.indexOf(memberName);` done for any particular reason? is it faster than a comparison? It's definitely got me confused as to what it's trying to do

Comment (treshugart) [inline src/index.js:852]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_NOT

It is faster (in micro-benchmarks) but is a little obtuse.

Comment (lukebatchelor) [inline src/index.js:852]: Oh, seems like its a reasonably common expression.

File: src/v0/index.js [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+  const createElementNS = document.createElementNS.bind(document);
+  document.createElement = (name, ...args) => createElement(filterTagName(name), ...args);
+  document.createElementNS = (name, ...args) => createElementNS(filterTagName(name), ...args);
+
+  // Override innerHTML to turn slot nodes into content nodes.
+  function replaceSlotsWithContents(node) {
+    const tree = document.createTreeWalker(node, NodeFilter.SHOW_ELEMENT);
+    const repl = [];
+
+    // Walk the tree and record nods that need replacing.
```
Comment (lukebatchelor) [inline src/v0/index.js]: Walk the tree and record **nods** that need replacing.

nodes

Comment (treshugart) [inline src/v0/index.js]: (nods)

**Comment by lukebatchelor** [comment]
lgtm
